### PR TITLE
NIFI-6576 add basic auth to confluent schema registry service

### PR DESCRIPTION
**Pull request by mark-weghorst**
#### Description of PR

_Adds basic authentication support to the Confluent schema registry controller; fixes bug #4224_

A pull request (https://github.com/apache/nifi/pull/4224) was originally created by @teeberg to resolve this issue, and @alopresto provided code review feedback on 27-Apr-2020, however no additional commits were provided to address the code review feedback.

This PR updates the code contributed in https://github.com/apache/nifi/pull/4224 to incorporate the code review changes requested in the original PR.

Current versions of the Confluent Schema Registry only support basic authentication, so support for digest authentication which was present in the original PR has been removed.  

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (alopresto) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:83]: Is there any requirement that the password is populated? If the `authType` is `BASIC` but the `authUser` is empty, basic auth won't be enabled but no error/warning will be communicated to the user.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (alopresto) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:207]: A password value of `    ` (4 space characters) for example would register as `isSet()` here but `StringUtils.trimToEmpty()` on line 165 above would return an empty string.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (alopresto) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:140]: I don't think setting the allowable value here to a static string is sufficient to enforce that the provided value will always be provided in all uppercase (via API call or manipulated submission), and there is no default value to offer an absence of authentication.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:83]: You raise an interesting question, AFAIK basic HTTP authentication requires a username, but will work with a password that is an empty string.  

A quick search found bugs in other projects where they are failing to support null passwords.  **_From my point of view the question is really whether we should allow empty passwords or not._**  

In your review of the original PR in April, you suggested that  customValidate for the controller be enhanced to throw an error if auth type was basic and either username or password were unset.  

I've done as  you requested in that orignal PR, and the UX will display a validation error with a meaningful message that both username and password are required when auth type is basic.  

Bottom line here is I'll bend to your judgement on this, let me know what you think we should do

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:207]: Good point, as i suggested in the other comment i think I need a decision on whether we want to support null passwords, and then i can rework this for you.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (alopresto) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:83]: Thanks Mark. I don't see anything in [RFC 7617](https://tools.ietf.org/html/rfc7617#section-2) which states that the username or password are actually required to be non-null. I think it depends on what the Confluent Schema Registry enforces in this case. As HTTP Basic Authentication provides no confidentiality of the credentials, we tend not to use it anywhere within NiFi, so when I made the recommendation I probably wasn't considering the use case where both the username and password are expected to be empty if CSR supports it. We also, to my knowledge, are not ensuring that the connection to CSR uses TLS if Basic Auth is enabled, in which case (often reused) credentials could be transmitted in cleartext. 

Personally I don't think we should introduce features to NiFi which allow the user to obliviously take actions which are insecure. If the CSR deployment accepts empty usernames/passwords, or accepts basic auth over HTTP, I generally think it's not a feature we should add to the platform, and users who are in scenarios that require this integration should be required to download/write/compile/deploy custom processors which we (the project community) are not responsible for as an explicit acknowledgment of this risk. I accept however, that doesn't always translate to the behaviors demonstrated in the real world. 

My suggestion then, is two-fold:

1. Check if the CSR URL is `http://` and if so, warn that basic auth credentials will be exposed (`WARN` level, not `ERROR` level)
1. Check if either the username or password is null/solely whitespace and warn that this is likely incorrect (but allow it if CSR allows it). This has the increased risk of potentially exposing to an observer that the password is empty, but if they are using basic auth with an empty password over plaintext HTTP, I don't know what we can do to help them at that point. 

From RFC 7617:

> 
>  To receive authorization, the client
> 
>    1.  obtains the user-id and password from the user,
> 
>    2.  constructs the user-pass by concatenating the user-id, a single
>        colon (":") character, and the password,
> 
>    3.  encodes the user-pass into an octet sequence (see below for a
>        discussion of character encoding schemes),
> 
>    4.  and obtains the basic-credentials by encoding this octet sequence
>        using Base64 ([RFC4648], Section 4) into a sequence of US-ASCII
>        characters ([RFC0020]).
> 
>    The original definition of this authentication scheme failed to
>    specify the character encoding scheme used to convert the user-pass
>    into an octet sequence.  In practice, most implementations chose
>    either a locale-specific encoding such as ISO-8859-1 ([ISO-8859-1]),
>    or UTF-8 ([RFC3629]).  For backwards compatibility reasons, this
>    specification continues to leave the default encoding undefined, as
>    long as it is compatible with US-ASCII (mapping any US-ASCII
>    character to a single octet matching the US-ASCII character code).
> 
>    The user-id and password MUST NOT contain any control characters (see
>    "CTL" in Appendix B.1 of [RFC5234]).
> 
>    Furthermore, a user-id containing a colon character is invalid, as
>    the first colon in a user-pass string separates user-id and password
>    from one another; text after the first colon is part of the password.
>    User-ids containing colons cannot be encoded in user-pass strings.

**Comment by pvillard31** [comment]
Hi @mark-weghorst and @alopresto - I'm interested by this additional feature.
Mark, do you think you're going to follow up on this (or are you waiting for additional feedback)?

**Comment by mark-weghorst** [comment]
@pvillard31 yes I'm still planning on re-working this to address the code review comments from @alopresto 

Unfortunately, I haven't had the time to revisit this due to some more pressing work, I should have some time towards the end of the week to pick back up on this.

**Comment by pvillard31** [comment]
awesome, let me know if I can be of any help!

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:209]: See comments on the Authentication Type property and recommend reviewing and integrating dependsOn property support if needed, which might remove the need for some of this custom validation.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:82]: If HTTP Basic is the only type of authentication supported, then recommend removing authType as a parameter and passing just the username and password.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:81]: Recommend changing these checks to use StringUtils.isNotEmpty() to include null checking and improve readability.

**Review by exceptionfactory — COMMENTED** [review]
@mark-weghorst Thanks for the contribution, I provided some additional comments and I would be glad to help get this feature to completion.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:83]: I went ahead and re-worked the validator to generate warnings when using schema registry without TLS encyption.  

As for the null password, while it is in fact RFC compliant the Confluent Schema registry does not list this as supported in their documentation.  

100% of the examples they give for configuration include a username/passsword combination.  So I've kept the requirement that both the username and password be set.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:207]: In the re-worked validator, I first trimToEmpty() and base my empty determination on the trimmed version.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:140]: Your point is well taken about API generated configurations bypassing the UI validation.  All comparisons will this value are now case insensitive.   I don't think we can really set a default value here, as the default is the absence of value (unset) unless .default("") == unset?

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:117]: Renamed as requested

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:119]: renamed as requested

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:125]: renamed as requested

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:127]: renamed as requested

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:134]: I'd like to leave this as is, since I think additional authentication types will be forthcoming in the future.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:138]: @pvillard31  can you help out with this one?

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:209]: @pvillard31 can you help with this one?

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:82]: See comments above about additional auth type forthcoming

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (mark-weghorst) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java:81]: Changed as requested

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/client/RestSchemaRegistryClient.java]: The [Confluent Server documentation](https://docs.confluent.io/platform/current/kafka-rest/production-deployment/confluent-server/security.html) mentions HTTP Basic and mutual TLS, as well as additional methods for SASL-based authentication between components.  Maintaining the parameter makes sense given that background.

Rather than relying on a particular string value, however, adding an AuthenticationType enum with the values of BASIC and NONE would help make the implementation clearer.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:140]: The [Confluent Server documentation](https://docs.confluent.io/platform/current/kafka-rest/production-deployment/confluent-server/security.html#http-basic-authentication) lists BASIC and NONE as options for the authentication method for certain clients.  Following that approach and having a default value of NONE, in combination with an AuthenticationType enum, would help make the configuration properties and implementation approach clearer.

**Comment by exceptionfactory** [comment]
@mark-weghorst Thanks for responding to comments and making updates.  Regarding the dependsOn properties, the current version of the Developer Guide has some helpful documentation under [Validating Processor Properties.](https://github.com/apache/nifi/blob/d773521ee01f3a34fdd21365a86c8de7b82077a6/nifi-docs/src/main/asciidoc/developer-guide.adoc#validating-processor-properties).  Following the example described, the `Username` and `Password` properties described could include a `dependsOn(PROP_AUTHENTICATION_TYPE, "BASIC")` builder method which would make them dependent on selecting `BASIC` as the Authentication Type property.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:134]: Leaving the property in place and adding an allowable value of _NONE_ sounds like a good approach.

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (pvillard31) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:140]:
```suggestion
        .allowableValues("BASIC", "NONE")
        .defaultValue("NONE")
```

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (pvillard31) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:122]:
```suggestion
        .addValidator(StandardValidators.createRegexMatchingValidator(Pattern.compile("^[\\x20-\\x39\\x3b-\\x7e\\x80-\\xff]+$")))
        .dependsOn(PROP_AUTH_TYPE, "BASIC")
```

File: nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java
Comment (pvillard31) [inline nifi-nar-bundles/nifi-confluent-platform-bundle/nifi-confluent-schema-registry-service/src/main/java/org/apache/nifi/confluent/schemaregistry/ConfluentSchemaRegistry.java:131]:
```suggestion
        .addValidator(StandardValidators.createRegexMatchingValidator(Pattern.compile("^[\\x20-\\x7e\\x80-\\xff]+$")))
        .dependsOn(PROP_AUTH_TYPE, "BASIC")
```

**Review by pvillard31 — CHANGES_REQUESTED** [review]
Hey @mark-weghorst - I made some suggestions in line with the review from @exceptionfactory, the suggestions are made from the Github web UI so I may have made mistakes but this should be close to what we'd like before merging this pull request to main. Let me know if you have any questions.

**Comment by KonstantinCodes** [comment]
This feature would be absolutely amazing. Thanks everyone for the work so far. Hope to use it soon :)

**Comment by pvillard31** [comment]
We pushed this over the finish line with @exceptionfactory, thanks for all the work on this PR @mark-weghorst !

**Comment by mark-weghorst** [comment]
Thank you @exceptionfactory  for your help getting this pushed over the finishing line over the holidays!

**Pull request closed without merging**
